src/main/scala/org/kframework/minikore/MiniKore.scala

```diff
@@ -1,39 +1,162 @@
 package org.kframework.minikore
 
+
+import org.kframework.minikore.TreeInterface._
+import org.kframework.minikore.{PatternInterface => i}
+
 import scala.collection._
 
 /** Algebraic data type of MiniKore. */
 object MiniKore {
 
-  type Attributes = Seq[Pattern]
+  //Default Implementation
+
+  type Attributes = Seq[i.Pattern]
 
   case class Definition(modules: Seq[Module], att: Attributes)
+
   case class Module(name: String, sentences: Seq[Sentence], att: Attributes)
 
   sealed trait Sentence
+
   case class Import(name: String, att: Attributes) extends Sentence
+
   case class SortDeclaration(sort: String, att: Attributes) extends Sentence
+
   case class SymbolDeclaration(sort: String, label: String, args: Seq[String], att: Attributes) extends Sentence
-  case class Rule(pattern: Pattern, att: Attributes) extends Sentence
-  case class Axiom(pattern: Pattern, att: Attributes) extends Sentence
-
-  sealed trait Pattern
-  case class Variable(name: String, sort: String) extends Pattern
-  case class Application(label: String, args: Seq[Pattern]) extends Pattern
-  case class DomainValue(label: String, value: String) extends Pattern
-  //
-  case class True() extends Pattern
-  case class False() extends Pattern
-  //
-  case class And(p: Pattern, q: Pattern) extends Pattern
-  case class Or(p: Pattern, q: Pattern) extends Pattern
-  case class Not(p: Pattern) extends Pattern
-  case class Implies(p: Pattern, q: Pattern) extends Pattern
-  case class Exists(v: Variable, p: Pattern) extends Pattern
-  case class ForAll(v: Variable, p: Pattern) extends Pattern
-  //
-  case class Next(p: Pattern) extends Pattern
-  case class Rewrite(p: Pattern, q: Pattern) extends Pattern
-  case class Equal(p: Pattern, q: Pattern) extends Pattern
+
+  case class Rule(pattern: i.Pattern, att: Attributes) extends Sentence
+
+  case class Axiom(pattern: i.Pattern, att: Attributes) extends Sentence
+
+  case class Variable(name: String, sort: i.Sort) extends i.Variable {
+    override def build = DefaultBuilders.VariableBuilder
+  }
+
+  case class Application(label: String, args: Seq[i.Pattern]) extends i.Application {
+    override def build = DefaultBuilders.ApplicationBuilder(label, _)
+  }
+
+
+  case class DomainValue(label: String, value: String) extends i.DomainValue {
+    override def build = DefaultBuilders.DomainValueBuilder
+  }
+
+  case class True() extends i.True {
```

Comment:
`Top` and `Bottom` is fine, we can do that change.

This discussion is reminding me that @hreada and I discussed that we may want to include the sorts of all patterns in their signature.  Here are the reasons:

1) Sorts are part of the semantics of ML, so all provers and backends should need them.

2) Frontends already compute the sorts of everything as part of their consistency checking, so why throw that information away and then have the backends re-infer it?

3) It is not nice to put that sort information in attributes, considering that ALL backends should need it.

Anybody can think about any argument against not putting the sort info in each pattern?